crates/extensions/tedge-p11-server/src/pkcs11.rs

```diff
@@ -383,3 +465,102 @@ fn write_asn1_integer(writer: &mut dyn std::io::Write, b: &[u8]) {
     let i = asn1_rs::Integer::new(&i);
     let _ = i.write_der(writer);
 }
+
+pub mod uri {
+    use std::borrow::Cow;
+    use std::collections::HashMap;
+
+    /// Attributes decoded from a PKCS #11 URL.
+    ///
+    /// Attributes only relevant to us shall be put into fields and the rest is in `other` hashmap.
+    ///
+    /// https://www.rfc-editor.org/rfc/rfc7512.html
+    #[derive(Debug, Clone)]
+    pub struct Pkcs11Uri<'a> {
+        pub token: Option<Cow<'a, str>>,
+        pub serial: Option<Cow<'a, str>>,
+        pub id: Option<Vec<u8>>,
+        pub object: Option<Cow<'a, str>>,
+        pub other: HashMap<&'a str, Cow<'a, str>>,
+    }
+
+    impl<'a> Pkcs11Uri<'a> {
+        pub fn parse(uri: &'a str) -> anyhow::Result<Self> {
+            let path = uri
+                .strip_prefix("pkcs11:")
+                .ok_or(anyhow::anyhow!("missing PKCS #11 URI scheme"))?;
+            // split of the query component
+            let path = path.split_once('?').map(|(l, _)| l).unwrap_or(path);
+
+            let pairs = path.split(';').filter_map(|pair| pair.split_once('='));
+
+            let mut pairs: HashMap<&str, &str> = HashMap::from_iter(pairs);
+
+            let token = pairs
+                .remove("token")
+                .map(|v| percent_encoding::percent_decode_str(v).decode_utf8_lossy());
+            let serial = pairs
+                .remove("serial")
+                .map(|v| percent_encoding::percent_decode_str(v).decode_utf8_lossy());
+            let object = pairs
+                .remove("object")
+                .map(|v| percent_encoding::percent_decode_str(v).decode_utf8_lossy());
+
+            let id: Option<Vec<u8>> = pairs
+                .remove("id")
+                .map(|id| percent_encoding::percent_decode_str(id).collect());
+
+            let other = pairs
+                .into_iter()
+                .map(|(k, v)| {
+                    (
+                        k,
+                        percent_encoding::percent_decode_str(v).decode_utf8_lossy(),
+                    )
+                })
+                .collect();
+
+            Ok(Self {
+                token,
+                serial,
+                id,
+                object,
+                other,
+            })
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        #[test]
+        fn decodes_pkcs11_uri() {
+            // https://www.rfc-editor.org/rfc/rfc7512.html#section-3
+            let input = "pkcs11:token=The%20Software%20PKCS%2311%20Softtoken;\
+            manufacturer=Snake%20Oil,%20Inc.;\
+            model=1.0;\
+            object=my-certificate;\
+            type=cert;\
+            id=%69%95%3E%5C%F4%BD%EC%91;\
+            serial=\
+            ?pin-source=file:/etc/token_pin";
+
+            // let input = "pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=5c1becd5222a84f6;token=token123;id=%01;object=object123;type=private";
```

Comment:
Is there a plan to test also this input?

Reply:
All of these attributes are already tested in the example above, so removed this comment in 0208319f.